litr/src/main/java/com/linkedin/android/litr/codec/PassthroughBufferEncoder.kt

```diff
@@ -0,0 +1,128 @@
+ */
+package com.linkedin.android.litr.codec
+
+import android.media.MediaCodec
+import android.media.MediaFormat
+import android.view.Surface
+import java.nio.ByteBuffer
+import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.LinkedBlockingQueue
+
+private const val DEFAULT_BUFFER_POOL_SIZE = 2
+
+/**
+ * Implementation of [Encoder] that passes input frames to output consumer without modifying them.
+ * Works only in ByteBuffer mode.
+ * @param inputBufferCapacity capacity of encoder's input buffers
+ */
+class PassthroughBufferEncoder(
+    inputBufferCapacity: Int
+) : Encoder {
+
+    // pool of unused frames
+    private val availableFrames = LinkedBlockingQueue<Frame>()
+    // pool of frames a client dequeued to populate with raw input data
+    private val dequeuedInputFrames = ConcurrentHashMap<Int, Frame>()
+    // queue of frames being encoded
+    private val encodeQueue = LinkedBlockingQueue<Int>()
+    // pool of encoded frames
+    private val encodedFrames = ConcurrentHashMap<Int, Frame>()
+
+    private lateinit var mediaFormat: MediaFormat
+
+    private var isRunning = false
+    private var mediaFormatChanged = false
+
+    init {
+        for (tag in 0 until DEFAULT_BUFFER_POOL_SIZE) {
+            availableFrames.add(Frame(tag, ByteBuffer.allocate(inputBufferCapacity), MediaCodec.BufferInfo()))
+        }
+    }
+
+    override fun init(targetFormat: MediaFormat) {
+        this.mediaFormat = targetFormat
+    }
+
+    override fun createInputSurface(): Surface? {
+        return null
+    }
+
+    override fun start() {
+        isRunning = true
+    }
+
+    override fun isRunning(): Boolean {
+        return isRunning
+    }
+
+    override fun signalEndOfInputStream() {}
+
+    override fun dequeueInputFrame(timeout: Long): Int {
+        // since this method modifies multiple data structures, we have to make it synchronous
+        synchronized(this) {
+            return availableFrames.firstOrNull()?.let { availableFrame ->
+                availableFrames.remove(availableFrame)
+                dequeuedInputFrames[availableFrame.tag] = availableFrame
+                availableFrame.tag
+            } ?: MediaCodec.INFO_TRY_AGAIN_LATER
+        }
+    }
+
+    override fun getInputFrame(tag: Int): Frame? {
+        return dequeuedInputFrames[tag]
+    }
+
+    override fun queueInputFrame(frame: Frame) {
+        // since this method modifies multiple data structures, we have to make it synchronous
+        synchronized(this) {
+            frame.buffer?.flip()
+            dequeuedInputFrames.remove(frame.tag)
+            encodeQueue.add(frame.tag)
+            encodedFrames[frame.tag] = frame
+        }
+    }
+
+    override fun dequeueOutputFrame(timeout: Long): Int {
+        if (!mediaFormatChanged) {
+            mediaFormatChanged = true
+            return MediaCodec.INFO_OUTPUT_FORMAT_CHANGED
+        }
+        return encodeQueue.poll() ?: MediaCodec.INFO_TRY_AGAIN_LATER
```

Comment:
Should we be passing the timeout to poll() here? I realize this is passthrough but are we sure no one's going to have a race condition queuing/dequeuing input/output frames?

Reply:
To clarify the previous comment a bit -- 
Because `timeout` is specified in the API, it's possible that someone might rely on it, so I'm wondering if it's something we should support even in this passthrough encoder.

Reply:
Good call. Added.